### PR TITLE
New API

### LINKED ISSUE #71: Performance degradation over time with ETS backend
**Describe the bug**
I've been benchmarking hot paths in https://github.com/plausible/analytics and noticed that ExHammer check [here](https://github.com/plausible/analytics/blob/af0b97e68ae8949a96743f07cee7ac4362fe353a/lib/plausible/site/gate_keeper.ex#L71) becomes slower over time, from 11us in the beginning of [k6](https://k6.io) benchmark (it [hits the same site](https://github.com/plausible/analytics/blob/af0b97e68ae8949a96743f07cee7ac4362fe353a/test/load/script.js#L6) every time, so `:hammer_ets_buckets` has a single key throughout the benchmark) to over 100ms in the end. And in some requests (when user-agents are cached) that `check_rate` call becomes the largest slice in the request flamechart

Screenshot 2023-11-22 at 19 28 43

** Provide the following details

- Elixir version (elixir -v): 1.15.7
- Erlang version (erl -v): 26.1.2
- Operating system: mac sonoma

### PR DESCRIPTION
**Pull request by ruslandoga**
This PR implements a new API https://github.com/ExHammer/hammer/pull/72#issuecomment-2094752030 inspired by https://github.com/michalmuskala/plug_attack, https://github.com/farhadi/rate_limiter, and Ecto.Repo.

Closes #71 

TODOs:
- ~~update Hammer.Plug~~ added documentation and updated CHANGELOG
- ~~write the upgrade guide~~ done
- ~~maybe add integration tests that can be used by all backends (similar to Ecto)?~~
- ~~add more ETS-based backends with different rate-limiting strategies (sliding window, leaky bucket, token bucket)?~~
- ~~add atomics-based backends?~~

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -2,56 +2,63 @@ name: CI
```
Comment (epinault) [inline .github/workflows/ci.yml:2]: why not creating a separate MR to update the pipeline and versions? , this would make this MR a little smaller

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -2,56 +2,63 @@ name: CI
...
+          otp-version: ${{matrix.otp}}
+          elixir-version: ${{matrix.elixir}}
 
       - uses: actions/cache@v4
         with:
           path: |
             deps
             _build
-          key: deps-${{ runner.os }}-${{ matrix.otp_version }}-${{ matrix.elixir_version }}-${{ hashFiles('**/mix.lock') }}
+          key: test-otp-${{ matrix.otp }}-elixir-${{ matrix.elixir }}-ref-${{ github.head_ref || github.ref }}-mix-${{ hashFiles('**/mix.lock') }}
```
Comment (epinault) [inline .github/workflows/ci.yml:44]: are you caching per commit?

File: .tool-versions [hunk 3]
```diff
@@ -1,2 +1,2 @@
-elixir 1.16.3-otp-26
```
Comment (epinault) [inline .tool-versions:1]: same for this . let s do it in a separate MR with pipeline change above

File: README.md [hunk 4]
```diff
@@ -1,85 +1,61 @@
...
+    {:hammer, "~> 7.0"}
   ]
 end
 ```
 
-## Documentation
+## Default Algorithm
 
-On HexDocs: [https://hexdocs.pm/hammer/frontpage.html](https://hexdocs.pm/hammer/frontpage.html)
+By default, Hammer uses a **fixed window counter** to track actions within set time windows, resetting the count at the start of each new window. For example, with a limit of 10 uploads per minute, a user could upload up to 10 files between 12:00:00 and 12:00:59, and up to 10 more between 12:01:00 and 12:01:59.
```
Comment (epinault) [inline README.md:25]: ❤️

File: README.md [hunk 5]
```diff
@@ -1,85 +1,61 @@
...
-config :hammer,
-  backend: {Hammer.Backend.ETS, [expiry_ms: 60_000 * 60 * 4,
-                                 cleanup_interval_ms: 60_000 * 10]}
+case MyApp.RateLimit.check_rate(key, scale, limit) do
+  {:allow, _count} -> :upload_the_video
+  {:deny, _limit} -> :deny_the_request
+end
 ```
 
-See the [Tutorial](https://hexdocs.pm/hammer/tutorial.html) for more.
```
Comment (epinault) [inline README.md:69]: what s wrong with the tutorial link? we should keep it?

File: mix.lock [hunk 6]
```diff
@@ -1,15 +1,14 @@
...
-  "makeup_erlang": {:hex, :makeup_erlang, "0.1.5", "e0ff5a7c708dda34311f7522a8758e23bfcd7d8d8068dc312b5eb41c6fd76eba", [:mix], [{:makeup, "~> 1.0", [hex: :makeup, repo: "hexpm", optional: false]}], "hexpm", "94d2e986428585a21516d7d7149781480013c56e30c6a233534bedf38867a59a"},
+  "credo": {:hex, :credo, "1.7.10", "6e64fe59be8da5e30a1b96273b247b5cf1cc9e336b5fd66302a64b25749ad44d", [:mix], [{:bunt, "~> 0.2.1 or ~> 1.0", [hex: :bunt, repo: "hexpm", optional: false]}, {:file_system, "~> 0.2 or ~> 1.0", [hex: :file_system, repo: "hexpm", optional: false]}, {:jason, "~> 1.0", [hex: :jason, repo: "hexpm", optional: false]}], "hexpm", "71fbc9a6b8be21d993deca85bf151df023a3097b01e09a2809d460348561d8cd"},
+  "dialyxir": {:hex, :dialyxir, "1.4.4", "fb3ce8741edeaea59c9ae84d5cec75da00fa89fe401c72d6e047d11a61f65f70", [:mix], [{:erlex, ">= 0.2.7", [hex: :erlex, repo: "hexpm", optional: false]}], "hexpm", "cd6111e8017ccd563e65621a4d9a4a1c5cd333df30cebc7face8029cacb4eff6"},
```
Comment (epinault) [inline mix.lock:4]: that version of 1.4.4 is problematic with OTP27. let s hold off till 1.4.5 is out

File: config/config.exs [hunk 7]
```diff
@@ -1,12 +0,0 @@
-# This file is responsible for configuring your application
-# and its dependencies with the aid of the Mix.Config module.
-import Config
-
-config :hammer,
-  backend:
-    {Hammer.Backend.ETS,
-     [
-       ets_table_name: :hammer_backend_ets_buckets,
-       expiry_ms: 60_000 * 60 * 2,
```
Comment (epinault) [inline config/config.exs:10]: what happened to this options?

File: guides/CreatingBackends.md [hunk 8]
```diff
@@ -1,79 +0,0 @@
-# Creating Backends
```
Comment (epinault) [inline guides/CreatingBackends.md:1]: why removing this ? we need a way let people know how to use the backend even with your new API? or is this moved somewhere else I missed?

File: guides/distributed-ets.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
+# Distributed Rate Limiter with ETS Backend
+
+The example implements a distributed, eventually consistent rate limiter using Phoenix.PubSub for broadcasting each hit across nodes and a local ETS backend to manage rate-limiting counters. This setup is useful when you need to limit the number of actions (e.g., requests) across multiple nodes in a cluster.
```
Comment (epinault) [inline guides/distributed-ets.md:3]: why not make it backend that is usable then by default in a separate MR?

File: guides/upgrade-v7.md [hunk 10]
```diff
@@ -0,0 +1 @@
+# Upgrading to Hammer V7
```
Comment (epinault) [inline guides/upgrade-v7.md:1]: still need instructions here?

File: lib/hammer.ex [hunk 11]
```diff
@@ -1,264 +1,60 @@
...
-  Example:
+      # this allows :ets to be aliased to Hammer.ETS
+      backend = with :ets <- backend, do: Hammer.ETS
+      @before_compile backend
 
-      user_id = 2406
-      {:ok, _count} = delete_buckets("file_uploads:\#{user_id}")
-
-  """
-  def delete_buckets(id) do
```
Comment (epinault) [inline lib/hammer.ex:189]: why can t we have this anynore?

File: lib/hammer/backend.ex [hunk 12]
```diff
@@ -1,41 +0,0 @@
-defmodule Hammer.Backend do
-  @moduledoc """
```
Comment (epinault) [inline lib/hammer/backend.ex:2]: why removing the behavior? we need one for other backedn type to ensure they implement they proper contract?

File: lib/hammer/ets.ex [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+        }
+      end
+
+      def start_link(opts) do
+        opts = Keyword.put(opts, :table, @table)
+        opts = Keyword.put_new(opts, :clean_period, :timer.minutes(10))
+        Hammer.ETS.start_link(opts)
+      end
+
+      def hit(key, scale, increment \\ 1) do
```
Comment (epinault) [inline lib/hammer/ets.ex:41]: I think we should document that API for other backend

File: mix.exs [hunk 14]
```diff
@@ -20,34 +19,27 @@ defmodule Hammer.Mixfile do
...
-    [mod: {Hammer.Application, []}, extra_applications: [:logger, :runtime_tools]]
+    [extra_applications: [:logger]]
   end
 
   defp deps do
     [
       {:credo, "~> 1.7", only: [:dev, :test]},
       {:ex_doc, "~> 0.30", only: :dev},
-      {:dialyxir, "~> 1.3", only: [:dev, :test], runtime: false},
-      {:poolboy, "~> 1.5"}
```
Comment (epinault) [inline mix.exs:32]: are you gonna let each backend handle their own pool?

File: lib/hammer/ets.ex [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+    full_key = {key, window}
+
+    case :ets.lookup(table, full_key) do
+      [{_full_key, count, _expires_at}] -> count
+      [] -> 0
+    end
+  end
+
+  @doc false
+  def wait(scale) do
```
Comment (epinault) [inline lib/hammer/ets.ex:104]: feels strange to have a wait in the API that is mostly use for testing?

File: test/hammer/ets_test.exs [hunk 16]
```diff
@@ -0,0 +1,153 @@
+defmodule Hammer.ETSTest do
+  use ExUnit.Case, async: true
+
+  defmodule RateLimit do
+    use Hammer, backend: :ets
```
Comment (epinault) [inline test/hammer/ets_test.exs:5]: let defined the wait function here instead?

**Comment by epinault** [comment]
Thank you for the contributions. Overall I like the new direction of instantiating a backend allowing you to use multiple backend in parallel . API looks good but I left some comments and concerns. Before releasing 7.0, I would want also fix at a minimum the redis backend along. As most people use that one. I also like the idea of a distributed ETS, so why not shipping this as separate MR? and then we could have one with Atomic as well separately? I could also see after that a different algorithm for window bucket

File: .github/workflows/ci.yml (hunk 2 above)
Comment (ruslandoga) [reply .github/workflows/ci.yml:44]: Per branch / PR.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (ruslandoga) [reply .github/workflows/ci.yml:2]: Separate PRs didn't work out the last time since it takes more time and effort ... I thought it would be easier this time to one-shot it :)

Comment (ruslandoga) [reply .github/workflows/ci.yml:2]: We can still split it all up if you'd like in the end.

File: .tool-versions (hunk 3 above)
Comment (ruslandoga) [reply .tool-versions:1]: I suggest removing this file, it doesn't bring any value and makes developers install new versions or update this lock each time. It's the job of CI, in my opinion, to ensure the source code works with the supported Elixir and Erlang versions.

File: config/config.exs (hunk 7 above)
Comment (ruslandoga) [reply config/config.exs:10]: In `:ets` it's now called `:clean_period`. It's passed into `RateLimit.start_link`.

File: guides/CreatingBackends.md (hunk 8 above)
Comment (ruslandoga) [reply guides/CreatingBackends.md:1]: Yes, it will be documented in the behaviors, similar to DBConnection.

File: guides/distributed-ets.md (hunk 9 above)
Comment (ruslandoga) [reply guides/distributed-ets.md:3]: I think there too many unknowns to make it a backend. There might be different ways people might want to use it or have different assumptions about how it works, and this way it's clear because the code is shown up front and it's up to them to adapt it to their needs.

File: guides/upgrade-v7.md (hunk 10 above)
Comment (ruslandoga) [reply guides/upgrade-v7.md:1]: Yes.

File: lib/hammer.ex (hunk 11 above)
Comment (ruslandoga) [reply lib/hammer.ex:189]: Confusing name, bad performance. Instead, there is `MyApp.RateLimit.reset(key, scale)` now.

File: lib/hammer/backend.ex (hunk 12 above)
Comment (ruslandoga) [reply lib/hammer/backend.ex:2]: I moved it into `Hammer` for now, but might move it into a separate module.

File: mix.exs (hunk 14 above)
Comment (ruslandoga) [reply mix.exs:32]: Yes. And ETS doesn't have a pool.

**Comment by ruslandoga** [comment]
I've updated the Redis backend: https://github.com/ExHammer/hammer-backend-redis/pull/56

**Comment by ruslandoga** [comment]
I'm not sure about the Plug. I'll try to update it but right now I don't actually seem much benefit from it.

Like, why is it better than

```elixir
plug :rate_limit_videos when action in :...

defp rate_limit_videos(conn, _opts) do
  user_id = get_session(conn, "user_id")

  case MyApp.RateLimit.hit("videos:#{user_id}", :timer.minutes(1), 10) do
    {:allow, _count} -> conn
    {:deny, _wait} -> conn |> send_resp(429, []) |> halt()
  end
end 
```

It doesn't even set any useful headers on the response.

File: README.md (hunk 5 above)
Comment (ruslandoga) [reply README.md:69]: I think tutorial can be part of the README, no need to make users look for it, click the links.

File: mix.lock (hunk 6 above)
Comment (ruslandoga) [reply mix.lock:4]: Doesn't seem to affect Hammer.

```console
$ mix dialyzer
Compiling 2 files (.ex)
Finding suitable PLTs
Checking PLT...
[:bunt, :compiler, :credo, :crypto, :earmark_parser, :eex, :elixir, :ex_doc, :ex_unit, :file_system, :hammer, :inets, :jason, :kernel, :logger, :makeup, :makeup_elixir, :makeup_erlang, :nimble_parsec, :stdlib]
Looking up modules in dialyxir_erlang-27.1.2_elixir-1.17.3_deps-dev.plt
Looking up modules in dialyxir_erlang-27.1.2_elixir-1.17.3.plt
Looking up modules in dialyxir_erlang-27.1.2.plt
Finding applications for dialyxir_erlang-27.1.2.plt
Finding modules for dialyxir_erlang-27.1.2.plt
Creating dialyxir_erlang-27.1.2.plt
Looking up modules in dialyxir_erlang-27.1.2.plt
Removing 3 modules from dialyxir_erlang-27.1.2.plt
Checking 19 modules in dialyxir_erlang-27.1.2.plt
Adding 196 modules to dialyxir_erlang-27.1.2.plt
done in 0m17.77s
Finding applications for dialyxir_erlang-27.1.2_elixir-1.17.3.plt
Finding modules for dialyxir_erlang-27.1.2_elixir-1.17.3.plt
Copying dialyxir_erlang-27.1.2.plt to dialyxir_erlang-27.1.2_elixir-1.17.3.plt
Looking up modules in dialyxir_erlang-27.1.2_elixir-1.17.3.plt
Checking 215 modules in dialyxir_erlang-27.1.2_elixir-1.17.3.plt
Adding 258 modules to dialyxir_erlang-27.1.2_elixir-1.17.3.plt
done in 0m10.86s
Finding applications for dialyxir_erlang-27.1.2_elixir-1.17.3_deps-dev.plt
Finding modules for dialyxir_erlang-27.1.2_elixir-1.17.3_deps-dev.plt
Copying dialyxir_erlang-27.1.2_elixir-1.17.3.plt to dialyxir_erlang-27.1.2_elixir-1.17.3_deps-dev.plt
Looking up modules in dialyxir_erlang-27.1.2_elixir-1.17.3_deps-dev.plt
Checking 473 modules in dialyxir_erlang-27.1.2_elixir-1.17.3_deps-dev.plt
Adding 595 modules to dialyxir_erlang-27.1.2_elixir-1.17.3_deps-dev.plt
done in 0m47.39s
No :ignore_warnings opt specified in mix.exs and default does not exist.

Starting Dialyzer
[
  check_plt: false,
  init_plt: ~c"/Users/x/Developer/hammer/_build/dev/dialyxir_erlang-27.1.2_elixir-1.17.3_deps-dev.plt",
  files: [~c"/Users/x/Developer/hammer/_build/dev/lib/hammer/ebin/Elixir.Hammer.ETS.beam",
   ~c"/Users/x/Developer/hammer/_build/dev/lib/hammer/ebin/Elixir.Hammer.beam"],
  warnings: [:unknown]
]
Total errors: 0, Skipped: 0, Unnecessary Skips: 0
done in 0m1.51s
done (passed successfully)
```

File: .tool-versions (hunk 3 above)
Comment (epinault) [reply .tool-versions:1]: this is useful for ASDF and provide information of the developping version quickly

File: .github/workflows/ci.yml (hunk 1 above)
Comment (epinault) [reply .github/workflows/ci.yml:2]: thats fine. let s just go with it for now

File: README.md (hunk 5 above)
Comment (epinault) [reply README.md:69]: sure.

File: guides/distributed-ets.md (hunk 9 above)
Comment (epinault) [reply guides/distributed-ets.md]: fair. we can make it later but it would still be nice to have a default opinion that works so people can use it and then they can copy it or we can make it more configurable over time. Fine for now to document it. I can add i later the way I want it by default

**Comment by epinault** [comment]
> I'm not sure about the Plug. I'll try to update it but right now I don't actually see much benefit from it.
> 
> Like, why is it better than
> 
> ```elixir
> plug :rate_limit_videos when action in ...
> 
> defp rate_limit_videos(conn, _opts) do
>   user_id = conn.assigns.current_user.id
>   key = "videos:#{user_id}"
>   scale = :timer.minutes(1)
>   limit = 10
> 
>   case MyApp.RateLimit.hit(key, scale, limit) do
>     {:allow, _count} ->
>       conn
> 
>     {:deny, retry_after} ->
>       conn
>       |> put_resp_header("retry-after", Integer.to_string(div(retry_after, 1000)))
>       |> send_resp(429, [])
>       |> halt()
>   end
> end 
> ```
> 
> It doesn't even set any useful headers on the response. Just adds a confusing (IMO) abstraction. If it was up to me (it's not, I know) I would probably deprecate it and replace it with a guide :)

I am game to remove it. it might be useful for some people to have an overall throtling but then we could provide them a guide to do so instead for now

File: lib/hammer/ets.ex (hunk 15 above)
Comment (epinault) [reply lib/hammer/ets.ex]: so you would have a user request wait? seems like a risky behavior to hog up the web server personally but sure we  can keep it

**Comment by ruslandoga** [comment]
> it might be useful for some people to have an overall throtling

I think it's not that hard to add it to the endpoint module. Or the main router.

```elixir
defmodule MyAppWeb.Endpoint do
  use Phoenix.Endpoint
  
  plug RemoteIP
  plug :rate_limit
  
  # ...
  
  defp rate_limit(conn, _opts) do
    key = "web_requests:#{:inet.ntoa(conn.remote_ip)}"
    scale = :timer.minutes(1)
    limit = 1000
    
    case MyApp.RateLimit.hit(key, scale, limit) do
      {:allow, _count} ->
        conn

      {:deny, retry_after} ->
        retry_after_seconds = div(retry_after, 1000)

        conn
        |> put_resp_header("retry-after", Integer.to_string(retry_after_seconds))
        |> send_resp(429, [])
        |> halt()
    end
  end
end
```

---

I think we can add a deprecation warning now, in `init/1`. It would called once, only during compilation (unless in development), and it can print a similar "custom" made plug based on the options.

**Comment by ruslandoga** [comment]
I've updated the Mnesia backend: https://github.com/ExHammer/hammer-backend-mnesia/pull/36

**Comment by epinault** [comment]
I think you have some conflicts to resolve

**Comment by ruslandoga** [comment]
I see that :) But I think I'll do it in the end. My plan right now is to figure out a good way to use Mnesia backend and document it (tomorrow), then open a PR deprecating Hammer.Plug (also tomorrow), then clean everything up regarding the Backend behaviour (no ETA), and finally write an upgrade guide (no ETA).

File: .github/workflows/ci.yml [hunk 17]
```diff
@@ -2,56 +2,63 @@ name: CI
...
+
+  format:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: erlef/setup-beam@v1
+        with:
+          elixir-version: 1
+          otp-version: 27
+      - run: mix format --check-formatted
```
Comment (cliftonmcintosh) [inline .github/workflows/ci.yml:64]: Does the `with` block here work?

As I understand it, the `run:` key is used to specify a shell command or script to be executed, and it does not accept the `with:` field. The `with:` field is typically used in the context of actions that require input parameters, but not directly in a `run:` step.

I've tried to set up something similar with dialyzer and my editor warns me that I can't. ChatGPT also tells me this isn't allowed. So I'm curious if this works for you here.

Comment (ruslandoga) [reply .github/workflows/ci.yml:64]: 👋 @cliftonmcintosh 

I might be misunderstanding your question, but why would `with` be used here?

Comment (ruslandoga) [reply .github/workflows/ci.yml:64]: If your question is about the `with` on `uses: erlef/setup-beam@v1` then yes, it works.

Comment (cliftonmcintosh) [reply .github/workflows/ci.yml:64]: NVM. I see that it is valid because the with block is for the setup step

**Review by epinault — APPROVED** [review]

**Comment by ruslandoga** [comment]
I think this was merged prematurely :)

There are still unfinished documents and links to my fork.

**Comment by epinault** [comment]
I'll fix up these file separately or you can submit PR on those items . Separating would be nice to avoid the PR to linger too long too

**Comment by epinault** [comment]
looking at them while not perfect I am not seeing what s wrong with them . this is doc and can be easily fix up. I am going to release it

as for the upgrade guide, it showing in order what you need to do to upgrade. it s pretty clear I would think and they can easily use AI and the step as prompt instructions.  The better way would be to use igniter and provide and upgrade recipe to do it for them but would not be 100% prefect and require quite a bit of time to build

**Comment by ruslandoga** [comment]
@epinault 

As a contributor, I get at least the right to decide when my work is complete. Please don't make a release until I think it's ready :)

**Comment by ruslandoga** [comment]
Also please stop merging the PRs that have TODOs in them, even if you don't understand their meaning.

**Comment by ruslandoga** [comment]
If you don't think you can work that way -- which is standard for most of open source -- please revert all of my contributions.

**Comment by epinault** [comment]
@ruslandoga I appreciate your contributions, but I need this project to move forward. So This is why I am pushing forward. I have been planning on similar work earlier this year and could not get to it. Your MR motivated me to get it moving. I am not quite following what is the problem to keep it moving forward here?  

I have been doing OSS for a long time and frequently leave MR that the owner can choose to merge or not. I am confused what the issue is here?  OSS is about collaboration and I appreciate your collaboration but not getting why you are having an issue with this work?

As for the TODO, every project has different guideline and style, see the credo file I have committed which does not allow for TODO, I much prefer using issue to track down the work left as TODO get lost.

**Comment by epinault** [comment]
> If you don't think you can work that way -- which is standard for most of OSS -- please revert all of my contributions.
I can work with you and we can chat maybe to have a better understanding with each others on how to collaborate moving forward. Would you like me to still remove it?

**Comment by ruslandoga** [comment]
Thank you for your thoughtful response and your willingness to collaborate further. I appreciate the effort you've put into maintaining this project and your openness to working together.

That said, open-source contributions are something I do for personal satisfaction and the joy of completing tasks to my own standards. This experience hasn't aligned with those motivations, so I believe it's best for me to move on.

Please feel free to use the work I've already contributed, and I wish you and the project continued success.
